core/dispatch.c

```diff
@@ -874,6 +874,17 @@ dispatch_enter_dynamorio(dcontext_t *dcontext)
         if (get_at_syscall(dcontext))
             handle_post_system_call(dcontext);
 
+#ifdef X86
+        /* If the next basic block starts at a debug register value,
+         * we fire a single step exception before getting to the basic block. */
```

Comment:
You are right : it should be _before_ the instruction is executed.
I did not manage to make it work with the single step mechanism, because it executes one instruction in a basic block before crafting the single step exception
I will try to do it, replacing the instruction by a nop.